lib/key_manager.rb

```diff
@@ -0,0 +1,71 @@
+require 'securerandom'
+require 'thread'
+
+class KeyManager
+  attr_reader :keys, :blocked_keys
+
+  def initialize
+    @keys = {}
+    @blocked_keys = {}
+    @mutex = Mutex.new
+  end
+
+  def generate_key
+    key = SecureRandom.hex(16)
+    # 5 min expiry
+    expiry = Time.now + 300 
+    @mutex.synchronize do
+      @keys[key] = expiry
+    end
+    key
+  end
+
+  def get_available_key
+    @mutex.synchronize do
+      now = Time.now
+      available_keys = @keys.reject { |_, expiry| expiry < now }
+      unblocked_keys = available_keys.reject { |key, _| @blocked_keys.key?(key) }
+
+      return nil if unblocked_keys.empty?
+
+      key, _ = unblocked_keys.to_a.sample
+      # block this key for 60 seconds
+      @blocked_keys[key] = now + 60 
+      key
+    end
+  end
+
+  def unblock_key(key)
+    @mutex.synchronize do
+      return false unless @keys.key?(key)
+      @blocked_keys.delete(key)
+      @keys[key] = Time.now + 300
+      true
+    end
+  end
+
+  def delete_key(key)
+    @mutex.synchronize do
+      return false unless @keys.key?(key) 
```

Comment:
Is there a need to check this during key delete ?

Reply:
I thought of adding this check because user may enter an invalid key and deleting can't be performed.